sdk/include/opentelemetry/sdk/logs/logger_provider.h

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "opentelemetry/logs/logger_provider.h"
+#include "opentelemetry/logs/noop.h"
+#include "opentelemetry/nostd/shared_ptr.h"
+#include "opentelemetry/sdk/common/atomic_shared_ptr.h"
+#include "opentelemetry/sdk/logs/logger.h"
+#include "opentelemetry/sdk/logs/processor.h"
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace logs
+{
+class LoggerProvider final : public opentelemetry::logs::LoggerProvider
+{
+public:
+  /**
+   * Initialize a new logger provider with a specified processor
+   * @param processor The processor for this logger provider. This must
+   * not be a nullptr.
+   */
+  explicit LoggerProvider(std::shared_ptr<LogProcessor> processor) noexcept;
+
+  /**
+   * Initialize a new logger provider. A processor must later be assigned
+   * to this logger provider via the SetProcessor() method.
+   */
+  explicit LoggerProvider() noexcept;
+
+  /**
+   * Creates a logger with the given name, and returns a shared pointer to it.
+   * If a logger with that name already exists, return a shared pointer to it
+   * @param name The name of the logger to be created.
+   * @param options (OPTIONAL) The options for the logger. TODO: Give list of options
+   */
+  opentelemetry::nostd::shared_ptr<opentelemetry::logs::Logger> GetLogger(
+      opentelemetry::nostd::string_view name,
+      opentelemetry::nostd::string_view options = "") noexcept override;
+
+  /**
+   * Creates a logger with the given name, and returns a shared pointer to it.
+   * If a logger with that name already exists, return a shared pointer to it
+   * @param name The name of the logger to be created.
+   * @param args (OPTIONAL) The arguments for the logger.
+   */
+  opentelemetry::nostd::shared_ptr<opentelemetry::logs::Logger> GetLogger(
+      opentelemetry::nostd::string_view name,
+      nostd::span<nostd::string_view> args) noexcept override;
+
+  /**
+   * Removes a logger with the given name. If no logger exists with that name,
+   * return false.
+   * @param name The name of the logger to be removed.
+   */
+  bool RemoveLogger(std::string name) noexcept;
```

Comment:
Curious - why do we need this?

Reply:
I was thinking that if a user creates a logger instance in a scope, then after user exits the scope they would lose the logger reference. The LoggerProvider would still have the logger registered though, so the user would call the RemoveLogger(name) method to deregister it. Hmm now that I type that it sounds like it would belong in the logger destructor such that it's automatically done. Would this work?

Reply:
Incorrect call to this function can lead to existence of dangling logger(s) - i.e., loggers removed from `LoggerProvider` map, but still used by application for logging.

Reply:
Inside the `RemoveLogger` method we first convert the Logger reference from an SDK Logger into a no-op logger before removing it from the `LoggerProvider` map. So in this exact case the user could still have a pointer to the logger, but calling its `log` method would no longer do anything.

Reply:
Then the function doesn't sound like remove logger, probably `DisableLogger` is an better alternative?

Reply:
Not sure about this, but I think loggers once created (maybe based on some configuration) should stay available during the existence of the logger provider. `GetLogger` can called in different program parts (libraries), and removing the logger in one place might cause negative effects in other places.

I'd vote to remove this method, until we have a compelling use case to add it.